### ISSUE TITLE
Failure to bind AAR library that depends on AppCompat and uses camera API

### ISSUE THREAD

**Issue by DoDoENT** [xamarin/xamarin-android#1874]
### Steps to Reproduce

1. Create an AAR binding library as described [in official documentation](https://developer.xamarin.com/guides/android/advanced_topics/binding-a-java-library/binding-an-aar/), but instead of example AAR, use AAR that depends on AppCompat and uses camera API, like for example [BlinkID Android AAR](https://github.com/BlinkID/blinkid-android/blob/master/LibBlinkID.aar).
2. Try building the binding library
3. It fails with these errors: [xamarin.log](https://github.com/xamarin/xamarin-android/files/2124255/xamarin.log)

Reference AppCompat by adding package `Xamarin.Android.Support.v7.AppCompat` to the binding library, but this fails with following log:

```
Attempting to gather dependency information for package 'Xamarin.Android.Support.v7.AppCompat.27.0.2.1' with respect to project 'BlinkIDAARBinding', targeting 'MonoAndroid,Version=v4.1'
  GET https://api.nuget.org/v3/registration3-gz-semver2/xamarin.android.support.v7.appcompat/index.json
  OK https://api.nuget.org/v3/registration3-gz-semver2/xamarin.android.support.v7.appcompat/index.json 601ms
Total number of results gathered : 1
Gathering dependency information took 611.5 ms
Summary of time taken to gather dependencies per source :
https://api.nuget.org/v3/index.json	-	611.05 ms
Attempting to resolve dependencies for package 'Xamarin.Android.Support.v7.AppCompat.27.0.2.1' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
Resolving actions to install package 'Xamarin.Android.Support.v7.AppCompat.27.0.2.1'
Resolved actions to install package 'Xamarin.Android.Support.v7.AppCompat.27.0.2.1'
Retrieving package 'Xamarin.Android.Support.v7.AppCompat 27.0.2.1' from 'nuget.org'.
For adding package 'Xamarin.Android.Support.v7.AppCompat.27.0.2.1' to project 'BlinkIDAARBinding' that targets 'monoandroid41'.
Install failed. Rolling back...
Package 'Xamarin.Android.Support.v7.AppCompat.27.0.2.1' does not exist in project 'BlinkIDAARBinding'
Package 'Xamarin.Android.Support.v7.AppCompat.27.0.2.1' does not exist in folder '/Users/dodo/Work/Microblink/deployRepos/blinkid/blinkid-xamarin/Samples/BlinkID-Xamarin-Android/packages'
Executing nuget actions took 131.95 ms
Could not install package 'Xamarin.Android.Support.v7.AppCompat 27.0.2.1'. You are trying to install this package into a project that targets 'MonoAndroid,Version=v4.1', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.
```

Next thing I tried is to manually download [AppCompat AAR](https://maven.google.com/com/android/support/appcompat-v7/27.1.1/appcompat-v7-27.1.1.aar) and create another binding library for it, using [the same instructions](https://developer.xamarin.com/guides/android/advanced_topics/binding-a-java-library/binding-an-aar/) as for my original AAR, however when building this binding library, I get following errors in build log:

[xamarin-appcompat.log](https://github.com/xamarin/xamarin-android/files/2124293/xamarin-appcompat.log)

### Expected Behavior
The created binding library should build and work.

### Actual Behavior
See above.

### Version Information

Visual Studio Community 2017 for Mac
Version 7.5.1 (build 22)
Installation UUID: af0c96ed-b55d-45fa-b2f3-1fef37c9d8af
Runtime:
	Mono 5.10.1.47 (2017-12/8eb8f7d5e74) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 4.4.1.178 (master / eeaeb7e6)

	Package version: 510010047

NuGet
Version: 4.3.1.4445

.NET Core
Runtime: Not installed
SDK: Not installed
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.10.1/lib/mono/msbuild/15.0/bin/Sdks

Xamarin.Profiler
Version: 1.6.2
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Apple Developer Tools
Xcode 9.4.1 (14161)
Build 9F2000

Xamarin.Mac
Version: 2.10.0.103 (Visual Studio Community)

Xamarin.iOS
Version: 11.10.1.178 (Visual Studio Community)
Hash: 408d3574
Branch: d15-7
Build date: 2018-05-08 18:56:30-0400

Xamarin.Android
Version: 8.3.0.19 (Visual Studio Community)
Android SDK: /Users/dodo/android-sdks
	Supported Android versions:
		4.1 (API level 16)
		4.2 (API level 17)
		5.1 (API level 22)
		6.0 (API level 23)
		7.0 (API level 24)
		7.1 (API level 25)
		8.0 (API level 26)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.0
SDK Build Tools Version: 28.0.0 rc2

Java SDK: /usr
java version "1.8.0_162"
Java(TM) SE Runtime Environment (build 1.8.0_162-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.162-b12, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Xamarin Inspector
Version: 1.4.0
Hash: b3f92f9
Branch: master
Build date: Fri, 19 Jan 2018 22:00:34 GMT
Client compatibility: 1

Build Information
Release ID: 705010022
Git revision: 60442dd643a20c7a4ae1f8705b8d1de8972eee78
Build date: 2018-05-15 01:43:39+00
Xamarin addins: 4194ffe4868321e4c3477bd56aed579bda4c6fbb

Operating System
Mac OS X 10.13.5
Darwin 17.6.0 Darwin Kernel Version 17.6.0
    Tue May  8 15:22:16 PDT 2018
    root:xnu-4570.61.1~1/RELEASE_X86_64 x86_64

### Log File
Build logs attached above.

**Comment by jonathanpeppers** [comment]
@DoDoENT it looks like your binding project's `TargetFrameworkVersion` is set to `v4.1`. The NuGet package requires `v8.1`.

If you are using the support library 27.x, you should be using `TargetFrameworkVersion` `v8.1` for API 27.

On VS for Mac, this setting is under `Project Options | Build | General | Target framework`. Set it to `Android 8.1 (Oreo)`.

Can you check/close if that fixes your issue? thanks!

**Comment by DoDoENT** [comment]
Hi @jonathanpeppers.

After setting `TargetFrameworkVersion` to `v8.1` I was able to add `AppCompat` package to my binding project, but it still fails to compile as C# code generator generates code with duplicate fields and incorrectly handles interfaces. Here is the build log:

[xamarin.log](https://github.com/xamarin/xamarin-android/files/2124549/xamarin.log)

Is there a way to debug problems with C# code generator?

**Comment by jonathanpeppers** [comment]
@DoDoENT so unfortunately this is the norm when creating a binding project, most Java libraries of a reasonable size are going to require you to manually "fix things up" using `Metadata.xml`. Many of the issues are caused by the differences between C# and Java.

One such example in your log is:
```
error CS0534: 'SloveniaCombinedRecognizer.Result' does not implement inherited abstract member 'Entity.Result.Clone()'
```

This is a known issue, Java supports "covariant return types" and C# doesn't... `Entity.Result.Clone` returns `Entity.Result`, but `SloveniaCombinedRecognizer.Result.Clone` returns `SloveniaCombinedRecognizer.Result`, which is invalid in C#. I made an [attempt](https://github.com/xamarin/java.interop/pull/216) to make this situation work, but it turned out to be a *difficult* change.

So you have a couple approaches here:
1. Remove any `package` nodes you don't need to call from C#. Maybe you can remove all the broken ones? I have example removing packages [here](https://github.com/jonathanpeppers/glidex/blob/065f02664e8034207aa9cc2afb4e3f83f14c9e42/glidex/Transforms/Metadata.xml#L9-L17).
1. Add `Metadata.xml` fixups to make everything work.

For No. 2, I would recommend some reading.

Jon Douglas has a good guide on how to get started: https://gist.github.com/JonDouglas/dda6d8ace7d071b0e8cb

Also checkout the Xamarin documentation [here](https://docs.microsoft.com/en-us/xamarin/android/platform/binding-java-library/), specifically the topic about [covariant return types](https://docs.microsoft.com/en-us/xamarin/android/platform/binding-java-library/troubleshooting-bindings#problem-class-does-not-implement-interface-method).

Closing this for now as this has been the behavior for quite some time (sorry!). Feel free to reopen this or file a new issue if you find something that seems like a bug in the Java binding project system (it's not on the Xamarin troubleshooting documentation).

**Comment by DoDoENT** [comment]
From your [troubleshooting guide](https://developer.xamarin.com/guides/android/advanced_topics/binding-a-java-library/troubleshooting-bindings/) it appears that sometimes java obfuscation and ProGuard optimisation could cause problems to your C# code generator, so I created a debug version of my java AAR and tried to bind that, however I get the same build errors. Here is the log: 
[xamarin.log](https://github.com/xamarin/xamarin-android/files/2124604/xamarin.log)

Basically, the errors stating that some class does not implement inherited abstract member or interface member are the same I get with obfuscated version. However, in the obfuscated version I additionally get errors for duplicate fields and type mismatches, which could be attributed ProGuard's side-effects.

**Comment by DoDoENT** [comment]
Unfortunately, `Entity.Result.Clone` is part of the public API for my library, so I cannot remove it from the Java.

**Comment by jonathanpeppers** [comment]
If you have proguard-obfuscated types or member, those should be private/internal and hopefully not exposed to C#.

In some situations you may need to manually remove the types with a fixup in `Metadata.xml`.

Just like Java can change return types, it can also change visibility during inheritance...

**Comment by DoDoENT** [comment]
Obfuscated types are internal, but C# generator sees them, as they are inside the JAR.

Thanks for pointing me out to `Metadata.xml` - I'll investigate this. Basically, I would like to export to C# only API classes, and only their public members (private members of API classes can be obfuscated and should not be exposed to C#). Basically, instead of removing classes, I would like to specify what should remain (i.e. remove everything except stated), because the JAR will contain only a handful API classes - everything else is the implementation - and even the API classes will have internal stuff that does not need to be exposed, like implementation of the `android.os.Parcelable` and JNI bindings. 

Is there some more detailed documentation for writing `Metadata.xml` which I could read in order to achieve the described?

**Comment by jonathanpeppers** [comment]
@JonDouglas is this the best place to learn this: https://gist.github.com/JonDouglas/dda6d8ace7d071b0e8cb#2-fixing-issues

I think what Jon has is better than the Xamarin docs.

If `internal` types are getting bound to C#, it means they must be used during inheritance or something...

Like this class shouldn't get bound:
```java
//Omitted visibility is "package private" or "internal" in C#
class Foo {
    public void someMethod () { }
}
```

But if you added this one, `Foo` probably gets bound to C#:
```java
public class Bar extends Foo { }
```
Because C# will potentially need to be able to call the `someMethod` from `Foo` when using `Bar`...

If you find a simple situation where what we are doing seems *wrong*, file a bug, thanks!
